app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java

```diff
@@ -0,0 +1,233 @@
+                    if (!isValid) {
+                        errorList.add("Expected field " + fieldName + " to be " + fieldType.name() +
+                                ", but was instead JSON type " + fieldValueNode.getNodeType().name());
+                    }
+                } else if (isRequired) {
+                    errorList.add("Required field " + fieldName + " missing");
+                }
+            }
+        }
+
+        return errorList;
+    }
+
+    /**
+     * Given the expected field type and the field value JSON node, this validates the we have the correct type,
+     * returning true of the value is valid and false otherwise. Note that because attachment fields live in the
+     * attachments map and not in JSON, this will always return false for attachment types.
+     *
+     * @param fieldType
+     *         expected field type
+     * @param fieldValueNode
+     *         field value JSON node to validate
+     * @return true if valid, false otherwise
+     */
+    private static boolean validateJsonField(@Nonnull UploadFieldType fieldType, @Nonnull JsonNode fieldValueNode) {
+        switch (fieldType) {
+            case BOOLEAN:
+                return fieldValueNode.isBoolean();
+            case CALENDAR_DATE:
+                // We expect a string. Also, the string should be parseable by Joda LocalDate.
+                if (!fieldValueNode.isTextual()) {
+                    return false;
+                }
+
+                try {
+                    // DateUtils calls through to Joda parseLocalDate(), which is documented as
+                    // never returning null. So we don't need to null check here.
+                    DateUtils.parseCalendarDate(fieldValueNode.textValue());
+                    return true;
+                } catch (RuntimeException ex) {
+                    return false;
+                }
+            case FLOAT:
+            case INT:
+                // Research Kit doesn't distinguish between floats and ints, so neither should we. Just
+                // accept any number type.
+                return fieldValueNode.isNumber();
+            case INLINE_JSON_BLOB:
+                // JSON blobs are always JSON blobs. We don't need to do any special validation.
+                return true;
+            case STRING:
+                return fieldValueNode.isTextual();
+            case TIMESTAMP:
+                // either it's a string in ISO format, or it's a long in epoch milliseconds
+                if (fieldValueNode.isTextual()) {
+                    DateTime dateTime = UploadUtil.parseIosTimestamp(fieldValueNode.textValue());
+                    return dateTime != null;
+                } else if (fieldValueNode.isIntegralNumber()) {
+                    // any integral value can be converted to an epoch milliseconds, so this is good no
+                    // matter what
+                    return true;
+                } else {
```

Comment:
Unless the code is identical (like in FLOAT vs INT), I prefer not to use fall-through in switch statements.